src/transformStateWithClones.js

```diff
@@ -1,13 +1,31 @@
-'use strict';
-
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const stateHistory = [];
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(stateClone, obj.extraData);
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const delObj of obj.keysToRemove) {
+        delete stateClone[delObj];
+      }
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'clear') {
+      for (const key in stateClone) {
+        if (Object.prototype.hasOwnProperty.call(stateClone, key)) {
+          delete stateClone[key];
+        }
+      }
+      stateHistory.push({ ...stateClone });
+    }
```

Comment:
[CODE STYLE]: Use switch statement if you have a limited amount of conditions as it will make the code more readable and structured. You can replace the if statements with a switch statement that checks the `obj.type` value.